libstuff/SQValue.cpp

```diff
@@ -0,0 +1,136 @@
+        case TYPE::BLOB:
+            return text;
+        case TYPE::INTEGER:
+            return std::to_string(integer);
+        case TYPE::REAL:
+            // This is the same format that the sqlite3 shell uses internally.
+            char buf[64];
+            sqlite3_snprintf(sizeof(buf), buf, "%!.15g", real);
+            return string(buf);
+        case TYPE::NONE:
+        default:
+            return "";
+    }
+}
+
+string operator+(string lhs, const SQValue& rhs) {
+    lhs += static_cast<string>(rhs);
+    return lhs;
+}
+
+string operator+(const SQValue& lhs, const string& rhs) {
+    return static_cast<string>(lhs) + rhs;
+}
+
+string operator+(const char* lhs, const SQValue& rhs) {
+    return string(lhs) + static_cast<string>(rhs);
+}
+
+string operator+(const SQValue& lhs, const char* rhs) {
+    return static_cast<string>(lhs) + string(rhs);
+}
+
+string operator+(const SQValue& lhs, const SQValue& rhs) {
+    return static_cast<string>(lhs) + static_cast<string>(rhs);
+}
+
+bool operator==(const SQValue& a, const string& b) {
+    return static_cast<string>(a) == b;
+}
+
+bool operator==(const string& a, const SQValue& b) {
+    return a == static_cast<string>(b);
+}
+
+bool operator==(const SQValue& a, const char* b) {
+    return static_cast<string>(a) == string(b ? b : "");
+}
+
+bool operator==(const char* a, const SQValue& b) {
+    return string(a ? a : "") == static_cast<string>(b);
+}
+
+bool operator==(const SQValue& a, const SQValue& b) {
+    // Types must match
+    if (a.type != b.type) {
+        return false;
+    }
+
+    switch (a.type) {
+        case SQValue::TYPE::NONE:
+            // Consider two NONEs equal
+            return true;
+
+        case SQValue::TYPE::INTEGER:
+            return a.integer == b.integer;
+
+        case SQValue::TYPE::REAL:
+        {
+            // Floating point equality is fuzzy.
+            double aVal = a.real;
+            double bVal = b.real;
+            double difference = aVal - bVal;
+            if (difference < 0) {
+                difference = -difference;
+            }
+
+            const double absTol = 1e-12;
+            const double relTol = 1e-9;
+
+            double absA = aVal;
+            if (absA < 0) {
+                absA = -absA;
+            }
+            double absB = bVal;
+            if (absB < 0) {
+                absB = -absB;
+            }
+            double scaleFactor = (absA > absB) ? absA : absB;
+
+            return difference <= absTol + relTol * scaleFactor;
+        }
+
+        case SQValue::TYPE::TEXT:
+        case SQValue::TYPE::BLOB:
+            return a.text == b.text;
+    }
+
+    return false;
+}
+
+bool operator!=(const SQValue& a, const SQValue& b) {
+    return !(a == b);
+}
+
+bool SQValue::empty() const {
+    if (type == TYPE::NONE) {
+        return true;
+    }
+    if ((type == TYPE::TEXT || type == TYPE::BLOB) && text.empty()) {
+        return true;
+    }
+    return false;
```

Comment:
My understanding is the other types don't support `.empty()` and throw a compile error - so would this result cause any confusion or silent errors?

Reply:
This (and all of these functions really) only exist to make existing code that calls `string` functions on `SQResult`s work for the new `SQValue`s. Essentially, somewhere someone did a query can checked `.empty()` on the result, expecting it was a `string`.

The value of `integer` or `real` types will never be the empty string, it might be a value that stringifies to `0`, but they're never "empty" in the string sense.